ledger/trie/iterator.go

```diff
@@ -0,0 +1,138 @@
+	//       Therefore, by recalling `n`, we have recalled all ancestors of `p`.
+	//     - If `n` is the right child, we haven't already searched through all of `p`
+	//       descendents (as the `p.LeftChild` must have been searched before)
+	//       Therefore, by recalling `n`, we have recalled all ancestors of `p`
+	// Hence, it follows that the head of the stack always satisfies the
+	// Descendents-First-Relationship. As we search the trie in DFS manner, each
+	// node of the trie is recalled (once). Hence, the algorithm iterates all
+	// nodes of the MTrie while guaranteeing Descendents-First-Relationship.
+
+	// unprocessedRoot contains the trie's root before the first call of Next().
+	// Thereafter, it is set to nil (which prevents repeated iteration through the trie).
+	// This has the advantage, that we gracefully handle tries whose root node is nil.
+	unprocessedRoot Node
+	stack           []Node
+}
+
+// NewNodeIterator returns a node NodeIterator, which iterates through all nodes
+// comprising the MTrie. The Iterator guarantees a DESCENDANTS-FIRST-RELATIONSHIP in
+// the sequence of nodes it generates:
+//   * Consider the sequence of nodes, in the order they are generated by NodeIterator.
+//     Let `node[k]` denote the node with index `k` in this sequence.
+//   * Descendents-First-Relationship means that for any `node[k]`, all its descendents
+//     have indices strictly smaller than k in the iterator's sequence.
+// The Descendents-First-Relationship has the following important property:
+// When re-building the Trie from the sequence of nodes, one can build the trie on the fly,
+// as for each node, the children have been previously encountered.
+func NewNodeIterator(tr *Trie) *NodeIterator {
+	// for a Trie with height H (measured by number of edges), the longest possible path contains H+1 vertices
+	stackSize := ledger.NodeMaxHeight + 1
+	i := &NodeIterator{
+		stack: make([]Node, 0, stackSize),
+	}
+	i.unprocessedRoot = tr.RootNode()
+	return i
+}
+
+func (i *NodeIterator) Next() bool {
+	if i.unprocessedRoot != nil {
+		// initial call to Next() for a non-empty trie
+		i.dig(i.unprocessedRoot)
+		i.unprocessedRoot = nil
+		return true
+	}
+
+	// the current head of the stack, `n`, has been recalled
+	// we now inspect n's parent and dig into the parent's right child, if necessary
+	n := i.pop()
+	if len(i.stack) > 0 {
+		// If there are more elements on the stack, the next element on the stack is n's parent `p`.
+		// Before we can recall `p`, we need to dig into the parent's right child, if we haven't
+		// done so already. As we decent into the left child with priority, the only case where
+		// we still need to dig into the right child is, if n is p's left child.
+		parent := i.peek()
+		if parent.LeftChild() == n {
+			i.dig(parent.RightChild())
+		}
+		return true
+	}
+	return false // as len(i.stack) == 0, i.e. there are no more elements to recall
+}
+
+func (i *NodeIterator) Value() Node {
```

Comment:
Since the iterator relies on its stack to peek the right node value, how would that work when it's called with an empty stack?

Reply:
Ping @ricardomgoncalves

Reply:
I mean something like in the above function.

```go
        if i.unprocessedRoot != nil {
		// initial call to Next() for a non-empty trie
		i.dig(i.unprocessedRoot)
		i.unprocessedRoot = nil
		return true
	}
```

Reply:
Ah I see. Personally I prefer it the way it currently is, because I don't feel like the flexibility is necessary in this case, I'm fine with a rigid package that should be used one single way and no other, as long as it's clear and well documented. WDYT?

Reply:
Yeah, its fine.